proseco/catalog.py

```diff
@@ -0,0 +1,146 @@
+import traceback
+import numpy as np
+
+from astropy.table import Column
+
+from .core import ACACatalogTable
+from .guide import get_guide_catalog, GuideTable
+from .acq import get_acq_catalog, AcqTable
+from .fid import get_fid_catalog, FidTable
+
+
+def get_aca_catalog(obsid=0, *, raise_exc=False, **kwargs):
+    try:
+        aca = _get_aca_catalog(obsid, **kwargs)
+
+    except Exception:
+        if raise_exc:
+            raise
+
+        aca = ACACatalogTable.empty()  # Makes zero-length table with correct columns
+        aca.meta['exception'] = traceback.format_exc()
+
+        if aca.acqs is None:
+            aca.acqs = AcqTable.empty()
+        if aca.fids is None:
+            aca.fids = FidTable.empty()
+        if aca.guides is None:
+            aca.guides = GuideTable.empty()
+
+    return aca
+
+
+def _get_aca_catalog(obsid=0, att=None, man_angle=None, date=None,
+                     dither_acq=None, dither_guide=None,
+                     t_ccd_acq=None, t_ccd_guide=None,
+                     detector=None, sim_offset=None, focus_offset=None,
+                     n_guide=None, n_fid=None, n_acq=None,
+                     include_ids=None, include_halfws=None,
+                     print_log=False, raise_exc=False):
+
+    aca = ACACatalogTable()
+
+    # Put at least the obsid in top level meta for now
+    aca.meta['obsid'] = obsid
+
+    aca.acqs = get_acq_catalog(obsid=obsid, att=att, date=date, t_ccd=t_ccd_acq,
+                               dither=np.max(dither_acq), man_angle=man_angle,
+                               print_log=print_log)
+
+    aca.fids = get_fid_catalog(detector=detector, sim_offset=sim_offset,
+                               focus_offset=focus_offset, acqs=aca.acqs,
+                               stars=aca.acqs.meta['stars'], print_log=print_log)
+
+    aca.guides = get_guide_catalog(obsid=obsid, att=att, date=date, t_ccd=t_ccd_guide,
+                                   n_guide=n_guide, dither=dither_guide,
+                                   stars=aca.acqs.meta['stars'], print_log=print_log)
+
+    # Make a merged starcheck-like catalog.  Catch any errors at this point to avoid
+    # impacting operational work (call from Matlab).
```

Comment:
I still don't see a need for the merged table at all on the Matlab side, so maybe this should be a method on ACACatalogTable that we could call (optionally) if we want it on the SOT side.

Reply:
PS My slot-determination code was a port of the matlab code and seemed to work.  I don't think this is right:
``

 slot  idx      id    type  sz   p_acq    mag    maxmag   yang     zang    dim   res  halfw
int64 int64   int64   str3 str3 float64 float64 float64 float64  float64  int64 int64 int64
----- ----- --------- ---- ---- ------- ------- ------- -------- -------- ----- ----- -----
    0     1         1  FID  8x8   0.000    7.00    8.00   922.59 -1737.89     1     1    25
    1     2         5  FID  8x8   0.000    7.00    8.00 -1826.28   160.17     1     1    25
    2     3         6  FID  8x8   0.000    7.00    8.00   388.59   803.75     1     1    25
    3     4 157288312  BOT  6x6   0.739    8.73   10.23 -2143.88 -1717.81    20     1   160
    4     5 157689696  BOT  6x6   0.666    8.83   10.33  2445.48  -331.04    20     1   160
    5     6 157288520  GUI  6x6   0.000    8.81   10.31 -2394.12    45.51     1     1    25
    6     7 157692544  ACQ  6x6   0.685    9.02   10.52   148.06  1677.10    20     1   160
    7     8 157693576  ACQ  6x6   0.971    9.67   11.17  1744.25  1487.00    20     1   160
    0     9 157289008  ACQ  6x6   0.959    9.73   11.23   137.10 -2375.30    20     1    80
    1    10 157691632  ACQ  6x6   0.922    9.44   10.94  2296.85   167.54    20     1   160
    2    11 157288248  ACQ  6x6   0.979    9.61   11.11   -56.00 -1535.49    20     1   120
    3    12 157690680  ACQ  6x6   0.650    9.19   10.69  2444.36  1203.55    20     1   160
```

Which looks to me like slot assignment to a BOT and an ACQ in slot 3
Inputs
```
cat = proseco.catalog.get_aca_catalog(att=(20, 20, 30), man_angle=90, date='2018:001', dither_acq=(8,
    ...: 8), dither_guide=(8, 40), t_ccd_acq=-5, t_ccd_guide=-20, detector='ACIS-S', sim_offset=0, focus_offse
    ...: t=0, n_guide=3, n_fid=0, n_acq=8, print_log=True)
```

Reply:
Sorry, you're right.  I went the other way to get the slot and type assigned back to the original tables (which I like), and couldn't immediately wrap my head around how do that with the original.  Will fix.

Reply:
Agreed that Matlab doesn't need a merged table.  From the Matlab perspective the `aca` object is just a nice container that has the acqs, fids, and guides.  It doesn't care that there is a merged star catalog as well.

I do want the merged table to be in the pickle and to be handy for generating load products for validation testing and comparison with flight catalogs.  Do you see a problem with that?  I don't think there are really any downsides.

Reply:
Fixed the slot problem.  The new version is more complex but enforces consistency between the different tables (slots and entry ordering).  But of course I'm not sure if Matlab maintains the row ordering.  The goal here being that `aca.acqs[4]['slot']` actually matches the slots of the flight catalog for the 5th acquisition star.

Reply:
Consistency in slot and type for a given AGASC ID, and a goal being to have the merged catalog be exactly the same as what we see in starcheck.

Reply:
OK, still not sure what that really means (with regard to changes in the code from my prototype).  The lack of option for monitors is probably the only thing that will keep these matching in starcheck anyway, but of course small number of observations.